File: android/src/main/java/it/innove/NotifyBufferContainer.java
package it.innove;

import java.nio.ByteBuffer;

public class NotifyBufferContainer {
    public ByteBuffer items;

    public NotifyBufferContainer(int size) {
        this.items = ByteBuffer.allocate(size);
    }
    public void resetBuffer(){
        this.items.clear();
    }
    public byte[] put(byte[] value){
        byte[] toInsert = null;
        byte[] rest = null;
        
        if (value.length > this.items.remaining()) {
            int restLength = value.length - this.items.remaining();
            rest = new byte[restLength];
            toInsert = new byte[this.items.remaining()];
            System.arraycopy(value, 0, toInsert, 0, toInsert.length);
            System.arraycopy(value, toInsert.length, rest, 0, rest.length);
        } else {
            toInsert = value;
        }
        
        this.items.put(toInsert);

        return rest;
    }
    public boolean isBufferFull(){
        return this.items.remaining() == 0;
    }
    public int size() {
        return this.items.limit();
    }
    @Override 
    protected void finalize() throws Throwable {
        this.items = ByteBuffer.allocate(0);
    }
}

File: ios/NotifyBufferContainer.swift
import Foundation

class NotifyBufferContainer {
    public var items: Data
    public let capacity: Int

    public var count: Int {
        items.count
    }

    public var remaining: Int {
        capacity - items.count
    }

    public var isBufferFull: Bool {
        items.count >= capacity
    }

    init(size: Int) {
        self.capacity = size
        self.items = Data(capacity: size)
    }

    public func resetBuffer() {
        items.removeAll(keepingCapacity: true)
    }

    public func put(_ value: Data) -> Data {
        let remainingCapacity = self.remaining
        guard remainingCapacity > 0 else { return value }

        let restLength = value.count - remainingCapacity

        let toInsert: Data
        let rest: Data
        if restLength > 0 {
            toInsert = value.prefix(remainingCapacity)
            rest = value.suffix(restLength)
        } else {
            toInsert = value
            rest = Data()
        }
        items.append(toInsert)

        return rest
    }
}
